walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,84 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Arrays;
+import com.google.sps.TimeRange;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    List<String> allAttendees = new ArrayList<>();
+    allAttendees.addAll(request.getAttendees());
+    allAttendees.addAll(request.getOptionalAttendees());
+    Collection<TimeRange> optionsConsideringAllAttendees = getAvailableTimes(events, request, allAttendees);
+    if (optionsConsideringAllAttendees.size() > 0) {
+      return optionsConsideringAllAttendees;
+    } else {
+      return getAvailableTimes(events, request, request.getAttendees());
+    }
+  }
+
+  private Collection<TimeRange> getAvailableTimes(Collection<Event> events, MeetingRequest request, Collection<String> attendees) {
+    // Store the times of events that attendees are already attending and sort by start time 
+    ArrayList<TimeRange> unavailabilities = new ArrayList<>();
+    for (String attendee : attendees) {
+        for (Event event : events) {
+            for (String busyAttendee : event.getAttendees()) {
+                if (attendee.equals(busyAttendee)) {
+                    unavailabilities.add(event.getWhen()); 
+                }
+            }
+        }
+    }
```

Comment:
Small optional optimization here:
You could make attendees into a HashSet or similar. Then you wouldn't need the outermost loop, as in the innermost loop you could do something like `if (attendees.contains(busyAttendee))`. This would be faster because HashSet lookup is constant time.

Reply:
Thank you!